pytential/symbolic/primitives.py

```diff
@@ -1497,11 +1493,13 @@ def __init__(self, kernel, density,
             raise ValueError("invalid value (%s) of qbx_forced_limit"
                     % qbx_forced_limit)
 
-        kernel_arg_names = {
-                karg.loopy_arg.name
-                for karg in (
-                    kernel.get_args()
-                    + kernel.get_source_args())}
+        source_kernels = tuple(source_kernels)
+        densities = tuple(densities)
+        kernel_arg_names = set()
+
+        for kernel in source_kernels + (target_kernel,):
+            for karg in (kernel.get_args() + kernel.get_source_args()):
+                kernel_arg_names.add(karg.loopy_arg.name)
```

Comment:
Could you add a check that all the source and target kernels have the same base kernel? (And only the "right" kind of operator on each?)